File: java/src/main/java/com/example/urlshortener/orchestrator/agents/CodebaseAnalysisAgent.java
package com.example.urlshortener.orchestrator.agents;

import com.example.urlshortener.orchestrator.Node;
import com.example.urlshortener.orchestrator.SharedContext;

import java.io.IOException;
import java.io.UncheckedIOException;
import java.nio.file.*;
import java.util.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;
import java.util.stream.Stream;

/**
 * Brownfield codebase-impact-analysis agent: a real (if simple) static scan
 * of the existing app/ source tree to answer which files mention the target
 * concept, and which HTTP endpoints in UrlController.java would be
 * affected. Direct port of orchestrator/agents/codebase_analysis_agent.py.
 */
public final class CodebaseAnalysisAgent {
    private CodebaseAnalysisAgent() {}

    private static final Pattern ROUTE_RE =
            Pattern.compile("@(Get|Post|Put|Delete|Patch)Mapping\\(\"([^\"]*)\"\\)");

    public static Map<String, Object> analyze(SharedContext ctx, Node node, String keyword, Path projectRoot) {
        Path appDir = projectRoot.resolve("src/main/java/com/example/urlshortener/app");
        List<Map<String, Object>> impactedFiles = new ArrayList<>();
        try (Stream<Path> paths = Files.walk(appDir)) {
            List<Path> javaFiles = paths.filter(p -> p.toString().endsWith(".java")).sorted().toList();
            for (Path path : javaFiles) {
                String content = Files.readString(path);
                if (content.toLowerCase().contains(keyword.toLowerCase())) {
                    List<Integer> matchingLines = new ArrayList<>();
                    String[] lines = content.split("\n", -1);
                    for (int i = 0; i < lines.length; i++) {
                        if (lines[i].toLowerCase().contains(keyword.toLowerCase())) matchingLines.add(i + 1);
                    }
                    String relPath = "src/main/java/com/example/urlshortener/app/" + appDir.relativize(path).toString().replace('\\', '/');
                    impactedFiles.add(Map.of("file", relPath, "matching_lines", matchingLines));
                }
            }
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }

        Path controllerPath = appDir.resolve("UrlController.java");
        String controllerText;
        try {
            controllerText = Files.readString(controllerPath);
        } catch (IOException e) {
            throw new UncheckedIOException(e);
        }
        List<String> impactedEndpoints = new ArrayList<>();
        Matcher m = ROUTE_RE.matcher(controllerText);
        while (m.find()) {
            String route = m.group(2);
            if (route.toLowerCase().contains(keyword.toLowerCase())) {
                impactedEndpoints.add(m.group(1).toUpperCase() + " " + route);
            }
        }
        boolean controllerImpacted = impactedFiles.stream().anyMatch(f -> "src/main/java/com/example/urlshortener/app/UrlController.java".equals(f.get("file")));
        if (impactedEndpoints.isEmpty() && controllerImpacted) {
            Matcher m2 = ROUTE_RE.matcher(controllerText);
            while (m2.find()) impactedEndpoints.add(m2.group(1).toUpperCase() + " " + m2.group(2));
        }

        List<String> fileNames = impactedFiles.stream().map(f -> (String) f.get("file")).toList();
        String dependencyNotes = "'" + keyword + "' appears in " + impactedFiles.size() + " existing module(s): "
                + fileNames + ". Downstream consumers that must keep working: UrlController.java's HTTP layer, "
                + "ShortenerServiceTest.java's unit tests, and UrlControllerTest.java's integration tests -- any "
                + "signature change to ShortenerService.java has to stay backward compatible or all three need "
                + "updating together.";

        ctx.recordDecision(
                node.stage, node.id,
                "Static impact scan for '" + keyword + "' across app/: " + impactedFiles.size() + " file(s), "
                        + impactedEndpoints.size() + " endpoint(s) affected.",
                dependencyNotes
        );

        Map<String, Object> out = new LinkedHashMap<>();
        out.put("keyword", keyword);
        out.put("impacted_files", impactedFiles);
        out.put("impacted_endpoints", impactedEndpoints);
        out.put("dependency_notes", dependencyNotes);
        return out;
    }
}

File: orchestrator/agents/codebase_analysis_agent.py
"""Brownfield codebase-impact-analysis agent.

Performs a real (if simple) static scan of the existing `app/` package to
answer: which files mention the target concept, and which HTTP endpoints in
main.py would be affected. This is genuine codebase reasoning over the actual
source on disk, not a canned description -- run it against a different
keyword or a changed codebase and the output changes accordingly.
"""
import os
import re

ROUTE_RE = re.compile(r'@app\.(get|post|put|delete|patch)\("([^"]+)"')


def analyze(ctx, node, keyword: str, project_root: str) -> dict:
    app_dir = os.path.join(project_root, "app")
    impacted_files = []
    for fname in sorted(os.listdir(app_dir)):
        if not fname.endswith(".py"):
            continue
        path = os.path.join(app_dir, fname)
        text = open(path, encoding="utf-8").read()
        if keyword.lower() in text.lower():
            matching_lines = [i + 1 for i, line in enumerate(text.splitlines()) if keyword.lower() in line.lower()]
            impacted_files.append({"file": f"app/{fname}", "matching_lines": matching_lines})

    main_text = open(os.path.join(app_dir, "main.py"), encoding="utf-8").read()
    impacted_endpoints = [
        f"{method.upper()} {route}" for method, route in ROUTE_RE.findall(main_text) if keyword.lower() in route.lower()
    ]
    # An endpoint can also be impacted through shared logic even if the route
    # string itself doesn't contain the keyword (e.g. analytics logic lives in
    # shortener.py and is exposed via /api/urls/{code}/analytics). If main.py
    # itself mentions the keyword, treat every endpoint it defines as in scope.
    main_py_impacted = any(f["file"] == "app/main.py" for f in impacted_files)
    if not impacted_endpoints and main_py_impacted:
        impacted_endpoints = [f"{method.upper()} {route}" for method, route in ROUTE_RE.findall(main_text)]

    dependency_notes = (
        f"'{keyword}' appears in {len(impacted_files)} existing module(s): "
        f"{[f['file'] for f in impacted_files]}. Downstream consumers that must keep working: "
        f"main.py's HTTP layer, tests/test_shortener.py's unit tests, and tests/test_api.py's "
        f"integration tests -- any signature change to shortener.py functions has to stay backward "
        f"compatible or all three need updating together."
    )

    ctx.record_decision(
        stage=node.stage,
        node_id=node.id,
        summary=f"Static impact scan for '{keyword}' across app/: {len(impacted_files)} file(s), "
        f"{len(impacted_endpoints)} endpoint(s) affected.",
        rationale=dependency_notes,
    )

    return {
        "keyword": keyword,
        "impacted_files": impacted_files,
        "impacted_endpoints": impacted_endpoints,
        "dependency_notes": dependency_notes,
    }
